src/vsr/README.md

```diff
@@ -0,0 +1,189 @@
+
+## Protocol: Request/Start View
+
+### `request_start_view`
+
+A backup sends a `command=request_start_view` to the primary of a view when any of the following occur:
+
+  - the backup learns about a newer view via a `command=commit` message, or
+  - the backup learns about a newer view via a `command=prepare` message, or
+  - a replica recovers to `status=recovering_head`
+
+### `start_view`
+
+When a `status=normal` primary receives `command=request_start_view`, it replies with a `command=start_view`.
+`command=start_view` includes the view's current suffix — the headers of the latest messages in the view.
+
+Upon receiving a `start_view` for the new view, the backup installs the suffix, transitions to `status=normal`, and begins repair.
+
+## Protocol: Repair Journal
+
+`request_headers` and `headers` repair gaps or breaks in a replica's journal headers.
+Because the headers are repaired backwards (from the head) by hash-chaining, it is safe for both backups and transitioning primaries.
+
+## Protocol: Repair WAL
+
+In response to a `request_prepare`:
+
+- Reply the `command=prepare` with the requested prepare, if available.
+- Reply `command=nack_prepare` if the request origin is the primary of the ongoing view-change and we never received the prepare. (This enables the primary to truncate uncommitted messages and remain available).
+- Otherwise do not reply.
+
+Per [PAR's CTRL Protocol](https://www.usenix.org/system/files/conference/fast18/fast18-alagappan.pdf), we do not nack corrupt entries, since they _might_ be the prepare being requested.
+
+## Protocol: Client
+
+1. Client registers with cluster (`command=request operation=register`). (See also: [Protocol: Normal](#protocol-normal)).
+2. Client receives `command=reply operation=register` from the cluster. (If the cluster is at the maximum number of clients, it evicts the oldest).
+3. Repeat:
+    1. Send `command=request` to cluster.
+    2. If the client has been evicted, receive `command=eviction` from the cluster. (The client must re-register before sending more requests.)
+    3. If the client has not been evicted, receive `command=reply` from cluster.
+
+See also:
+
+  - [Integration: Client Session Lifecycle](../../docs/usage/integration/md#client-session-lifecycle)
+  - [Integration: Client Session Eviction](../../docs/usage/integration/md#client-session-eviction)
+
+## Protocol: Repair Grid (Backup)
+
+TODO (Unimplemented)
+
+## Protocol: Repair: State Transfer
+
+TODO (Unimplemented)
+
+## Protocol: Reconfiguration
+
+TODO (Unimplemented)
+
+# Quorums
+
+- The _commit quorum_ is the minimum number of replicas required to complete a commit.
+- The _view-change quorum_ is the minimum number of replicas required to complete a view-change.
+
+With the default configuration:
+
+|      **Replica Count** |   1 |  2 |  3 |  4 |  5 |  6 |
+| ---------------------: | --: | -: | -: | -: | -: | -: |
+|      **Commit Quorum** |   1 |  2 |  2 |  3 |  3 |  3 |
```

Comment:
Are we using "commit" and "replication" quorums interchangable, or is there a prefered version?

Reply:
I'll change it to "replication", I just wasn't thinking about it when I wrote this.